File: Daily Challenge/March 2025/Count the Number of Complete Components.cs
public class Solution {
  public int CountCompleteComponents(int n, int[][] edges) {
    var adjList = new List<int>[n];
    for (int i = 0; i < n; i++) {
      adjList[i] = new List<int>();
    }
    foreach (var edge in edges) {
      adjList[edge[0]].Add(edge[1]);
      adjList[edge[1]].Add(edge[0]);
    }

    var visited = new bool[n];
    var components = new List<int>();
    int completeComponets = 0;

    for (int i = 0; i < n; ++i) {
      if (!visited[i]) {
        GetConnetedComponents(i, adjList, visited, components);
        bool isComplete = true;

        foreach (var node in components) {
          if (adjList[node].Count < (components.Count - 1)) {
            isComplete = false;
            break;
          }
        }

        if (isComplete) {
          completeComponets++;
        }
        components.Clear();
      }
    }

    return completeComponets;
  }

  private void GetConnetedComponents(int node, List<int>[] adjList, bool[] visited, List<int> components) {
    if (visited[node]) {
      return;
    }

    visited[node] = true;
    components.Add(node);

    foreach (var neighbor in adjList[node]) {
      GetConnetedComponents(neighbor, adjList, visited, components);
    }
  }
}

File: Daily Challenge/March 2025/Count the Number of Complete Components.cpp
class Solution {
public:
  int countCompleteComponents(int n, vector<vector<int>>& edges) {
    vector<vector<int>> adjList(n);
    for (auto& edge : edges) {
      adjList[edge[0]].push_back(edge[1]);
      adjList[edge[1]].push_back(edge[0]);
    }

    vector<bool> visited(n, false);
    vector<int> componnets;

    int completeComponents = 0;
    for (int i = 0; i < n; ++i) {
      if (!visited[i]) {
        getConnectedComponet(i, adjList, visited, componnets);

        bool isCompleteComponent = true;
        for (auto& node : componnets) {
          if (adjList[node].size() < (componnets.size() - 1)) {
            isCompleteComponent = false;
            break;
          }
        }

        if (isCompleteComponent) {
          ++completeComponents;
        }

        componnets.clear();
      }
    }

    return completeComponents;
  }

private:
  void getConnectedComponet(int node, vector<vector<int>>& adjList, vector<bool>& visited, vector<int>& componnets) {
    visited[node] = true;
    componnets.push_back(node);

    for (auto& neighbor : adjList[node]) {
      if (!visited[neighbor]) {
        getConnectedComponet(neighbor, adjList, visited, componnets);
      }
    }
  }
};